### PR TITLE
.dockerignore ignore support. Fixes #9171

### LINKED ISSUE #9171: .dockerignore should a support including ignored content
If the only thing you want in your dockerfile is the Gemfile of a project it is tedious to use .dockerignore to exclude everything manually.

With .gitignore there is support for including things allowing you to do:

```
/*
!/Gemfile
!/Gemfile.lock
```
Having similar for dockerignore would be very nice.

more background: http://stackoverflow.com/questions/5533050/gitignore-exclude-folder-but-include-specific-subfolder

### PR DESCRIPTION
**Pull request by omeid**
Signed-off-by: omeid public@omeid.me

### CODE REVIEW

File: docs/sources/reference/commandline/cli.md [hunk 1]
```diff
@@ -569,7 +569,8 @@ If a file named `.dockerignore` exists in the root of `PATH` then it
 is interpreted as a newline-separated list of exclusion patterns.
 Exclusion patterns match files or directories relative to `PATH` that
 will be excluded from the context. Globbing is done using Go's
-[filepath.Match](http://golang.org/pkg/path/filepath#Match) rules.
+[filepath.Match](http://golang.org/pkg/path/filepath#Match) rules with
+the extension of allow for exclusion rules by appending a rule with `!`.
```
Comment (duglin) [inline docs/sources/reference/commandline/cli.md]: I think we may need an example or two for people to full understand this.

File: integration-cli/docker_cli_build_dockerignore_test.go [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	if _, err = buildImageFromContext(name, ctx, true); err != nil {
+		t.Fatal(err)
+	}
+	logDone("build - test .dockerignore whole dir with .*")
+}
+
+func TestBuildLineBreak(t *testing.T) {
```
Comment (duglin) [inline integration-cli/docker_cli_build_dockerignore_test.go]: why was this moved?

File: pkg/fileutils/fileutils.go [hunk 3]
```diff
@@ -1,26 +1,49 @@
...
+	log "github.com/Sirupsen/logrus"
 )
 
 // Matches returns true if relFilePath matches any of the patterns
+// and isn't excluded by any of the subsequent patterns.
 func Matches(relFilePath string, patterns []string) (bool, error) {
-	for _, exclude := range patterns {
-		matched, err := filepath.Match(exclude, relFilePath)
+
+	var matched = false
```
Comment (duglin) [inline pkg/fileutils/fileutils.go]: minor nit but if you want to be clear about the default just make it:
```
matched := false
```

File: integration-cli/docker_cli_build_dockerignore_test.go (hunk 2 above)
Comment (omeid) [inline integration-cli/docker_cli_build_dockerignore_test.go]: The file was over 5,8k lines, it made my text editor slow to the point of being unusable. Regardless a 5,8k line files is hard to manage I think it was reasonable to break them. If it is a big deal, I can put it back in one.

Comment (duglin) [inline integration-cli/docker_cli_build_dockerignore_test.go]: I have mixed feelings about splitting "build" into two files, but I was more interested in why this one was moved because not about dockerignore.

Comment (omeid) [inline integration-cli/docker_cli_build_dockerignore_test.go]: Oh, I just realised it. I will put this one back right away and follow up with other comments.

Comment (duglin) [inline integration-cli/docker_cli_build_dockerignore_test.go]: ok - just please squash your commits down to one.

File: pkg/fileutils/fileutils.go [hunk 4]
```diff
@@ -1,26 +1,49 @@
...
+
+	for _, pattern := range patterns {
+
+		var negative bool
+		if pattern != "" && pattern[0] == '!' {
+			negative = true
+			pattern = pattern[1:]
+		}
```
Comment (duglin) [inline pkg/fileutils/fileutils.go]: Does ! by itself mean anything?  From checking Go's Match() it looks like "" doesn't match anything so I would think "!" shouldn't either. I'm thinking it would be good to add:
```
if pattern == "" {
  continue
}
```
after the above if statement.

**Comment by duglin** [comment]
CI is failing because you need to run your code changes through gofmt... see in the janky output where it says:
```
---> Making bundle: validate-gofmt (in bundles/1.5.0-dev/validate-gofmt)
These files are not properly gofmt'd:
 - integration-cli/docker_cli_build_dockerignore_test.go
 - pkg/fileutils/fileutils.go

Please reformat the above files using "gofmt -s -w" and commit the result.
```

Also, please squash your commits.

File: integration-cli/docker_cli_build_dockerignore_test.go [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+		"Makefile":         "all:",
+		".git/HEAD":        "ref: foo",
+		"src/x.go":         "package main",
+		"src/_vendor/v.go": "package main",
+		".gitignore":       "",
+		"README.md":        "readme",
+		"LICENSE.md":       "must be included.",
+		"docs/useless.md":  "history",
+		"docs/config.json": "{ config: 1 }",
+		".dockerignore":    ".git\npkg\n.gitignore\nsrc/_vendor\n*.md\n!LICENSE.md\bdocs\ndocs\n!docs/config.json",
```
Comment (duglin) [inline integration-cli/docker_cli_build_dockerignore_test.go]: can you convert this to:
```
".dockerignore": `
.git
pkg
.gitignore
src/_vendor
!LICENSE.cd
...
`
```
so its easier to read/follow

File: integration-cli/docker_cli_build_dockerignore_test.go [hunk 6]
```diff
@@ -0,0 +1,254 @@
+package main
```
Comment (duglin) [inline integration-cli/docker_cli_build_dockerignore_test.go]: I've thought about it more... and to make it easier to focus just on what's "new", please move all of this back into docker_cli_build_test.go for now.  After this PR we can then decide how we want to split this file up.

File: pkg/fileutils/fileutils.go [hunk 7]
```diff
@@ -1,26 +1,53 @@
...
+
+		var negative bool
+		if pattern == "" {
+			continue
+		}
+
+		if pattern[0] == '!' {
+			negative = true
+			pattern = pattern[1:]
+		}
```
Comment (duglin) [inline pkg/fileutils/fileutils.go]: This will let ! pass through and I think you need to have it skip it. Unless ! by itself means something.

**Comment by buddhamagnet** [comment]
@omeid apologies for any offence caused.

**Comment by buddhamagnet** [comment]
@omeid I have three unit tests stacked up for the fileutils package. Happy for me to commit them?

File: pkg/fileutils/fileutils.go (hunk 7 above)
Comment (omeid) [inline pkg/fileutils/fileutils.go]: `!` would mean nothing. As in don't match anything. Should I keep it or is it confusing?

File: docs/sources/reference/commandline/cli.md [hunk 8]
```diff
@@ -611,17 +612,20 @@ is not recursive.
...
 `/somedir/temporary.txt` would be ignored. The second line `*/*/temp*`, will
 ignore files starting with name `temp` from any subdirectory that is two levels
 below the root directory. For example, the file `/somedir/subdir/temporary.txt`
-would get ignored in this case. The last line in the above example `temp?`
+would get ignored in this case. The third line in the above example `temp?`
 will ignore the files that match the pattern from the root directory.
 For example, the files `tempa`, `tempb` are ignored from the root directory.
 Currently there is no support for regular expressions. Formats
-like `[^temp*]` are ignored.
+like `[^temp*]` are ignored. the last two lines means to ignore every markdown
```
Comment (duglin) [inline docs/sources/reference/commandline/cli.md]: s/the last/The last/

Comment (duglin) [inline docs/sources/reference/commandline/cli.md]: I assume order matters, right?  So !LICENSE.md before *md would make the !LICENSE have no effect, right?

File: pkg/fileutils/fileutils.go [hunk 9]
```diff
@@ -1,26 +1,56 @@
...
+			continue
+		}
+
+		if pattern[0] == '!' {
+			if len(pattern) == 1 {
+				continue
+			}
+			negative = true
+			pattern = pattern[1:]
+		}
```
Comment (duglin) [inline pkg/fileutils/fileutils.go]: Yes I think ! all by itself is kind of weird because its like negating the entire .dockerignore file - so it feels like they're asking for inconsistent things.  I would either ignore it, or throw an error - I'd lean towards erroring since silently ignoring a bad situation will probably lead to more confusion.

File: docs/sources/reference/commandline/cli.md (hunk 8 above)
Comment (omeid) [inline docs/sources/reference/commandline/cli.md]: Yup, the rules are evaluated in order.

Comment (duglin) [inline docs/sources/reference/commandline/cli.md]: I think it would be good to state that in the docs and elaborate using the previous example by saying that if the order were reversed then it would have no effect.

File: pkg/fileutils/fileutils.go (hunk 9 above)
Comment (omeid) [inline pkg/fileutils/fileutils.go]: Right now, it is ignoring it, I could return an error. "Invalid rule: `!`"?

Comment (omeid) [inline pkg/fileutils/fileutils.go]: This is probably better handled at the .dockerignore "parser" function, maybe not?

**Comment by buddhamagnet** [comment]
@omeid hope you don't mind I tidied up some of the documentation and made some of the text changes requested.

**Comment by buddhamagnet** [comment]
@omeid if we don't talk have a great weekend and sorry for the noise I emitted earlier!

File: pkg/fileutils/fileutils.go (hunk 9 above)
Comment (duglin) [inline pkg/fileutils/fileutils.go]: This is a tough one.  A warning might be nice but within this func I don't think we access to the output stream. So, if I had to choose between silently ignoring it and stopping/erroring, I'd pick error and halt the build. I think skipping what is really an error case w/o telling the user is too close to lying for my tastes :-)

Comment (duglin) [inline pkg/fileutils/fileutils.go]: Whether we do it here or in ReadDockerIgnore doesn't matter much to me.

Comment (duglin) [inline pkg/fileutils/fileutils.go]: But, let's add a testcase for it no matter which direction we go so that we make sure we get the expected behavior

**Comment by omeid** [comment]
All is well. thanks for the clean up. :)

**Comment by duglin** [comment]
Where are we on this one?  I think we need:
- throw an error (stop build) if `!` is by itself in the .dockerignore
- add a test for `!` 
- squash commits down to one

**Comment by buddhamagnet** [comment]
I have added a failing test for the single ! in the dockerignore - will leave the implementation up to @omeid as I'm not sure how he's like to handle it.

**Comment by buddhamagnet** [comment]
A few places where we could handle this error come to mind:

In builder/evaluator.go:

```go
excludes, _ := utils.ReadDockerIgnore(filepath.Join(b.contextPath, ".dockerignore"))
	if rm, _ := fileutils.Matches(".dockerignore", excludes); rm == true {
		os.Remove(filepath.Join(b.contextPath, ".dockerignore"))
		b.context.(tarsum.BuilderContext).Remove(".dockerignore")
	}
	if rm, _ := fileutils.Matches(b.dockerfileName, excludes); rm == true {
		os.Remove(filepath.Join(b.contextPath, b.dockerfileName))
		b.context.(tarsum.BuilderContext).Remove(b.dockerfileName)
	}
```

**Comment by buddhamagnet** [comment]
@omeid where are we on this? All well?

**Comment by camilonova** [comment]
Please. This is needed so much.

**Comment by ikks** [comment]
That's a good feature to have, looking forward to see this committed

**Comment by buddhamagnet** [comment]
@omeid I'm happy to take this on if you don't have time or are otherwise unable to progress.

**Comment by buddhamagnet** [comment]
@omeid I've opened a new PR at https://github.com/docker/docker/pull/11823 - started a new feature branch fresh from master and copied these changes across. Hope that makes things easier. If this offends you I will happily close it.

**Comment by icecrime** [comment]
I haven't been involved in the PR from the beginning so sorry if I'm misunderstanding, but it seems that #12245 now replaces this one, so I'm closing. Thanks!
